File: permutation_test/permutation_test_conf_int.m
clc
clear
close all

% This script demonstrates a confidence interval for a location parameter
% by inverting a permutation test

%%

%sample size
n = 10;

%set random seed
rng(42);

%normal sample
X = randn(n, 1);

%level of significance
alpha = 0.05;

m0s = linspace(-2, 2, 200);

for k = 1:length(m0s)
    %null hypothesis
    m0 = m0s(k);
    
    %sample deviations from null
    Xd = X - m0;
    
    %test statistic
    d = sum(Xd);
    
    S = zeros(2^n, n);
    %binary representation of signs, code segment adapted from 'de2bi' function
    for i = 1:2^n                  % Cycle through each element of the input vector/matrix.
        j = 1;
        tmp = i - 1;
        while (j <= n) && (tmp > 0)     % Cycle through each digit.
            S(i, j) = rem(tmp, 2);      % Determine current digit.
            tmp = floor(tmp/2);
            j = j + 1;
        end
    end
    S = fliplr(S);
    S(S == 0) = -1;
    
    %permutation distribution
    T = zeros(2^n, 1);
    for i = 1:2^n
        T(i) = sum(S(i, :).*Xd');
    end
    
    %one-sided p-value (alternative greater than)
    p(k) = sum(T > d)/(2^n);
    %two-sided p-value
    p2(k) = (sum(T > abs(d)) + sum(T < -abs(d)))/(2^n);
    
end

%%

plot(m0s, p2)

%confidence interval
lb = min(m0s(p2 >= alpha));
ub = max(m0s(p2 >= alpha));

File: permutation_test/permutation_test_conf_int.R
rm(list=ls()) #clear environment variables
cat('\014') #clear console

# This script implements a permutation test and confidence interval for a location
# parameter

library(pracma)

perm_test2 <- function(X, m0){
  #returns the p-value of a two sided permutation test with null hypothesis
  #location parameter of m0
  
  #sample size
  n <- length(X)
  
  #sample deviations from null
  Xd <- X - m0
  
  #test statistic
  d <- sum(Xd)
  
  #binary representation {-1, +1} for signs
  S <- matrix(0, nrow=2^n, ncol=n)
  for (i in 1:(2^n)){
    j <- 1
    tmp <- i - 1
    while ((j <= n) && (tmp > 0)){
      S[i, j] <- rem(tmp, 2)
      tmp <- floor(tmp/2)
      j <- j + 1
    }
  }
  S <- fliplr(S)
  S[S == 0] <- -1
  
  #permutation distribution
  T <- matrix(nrow = 2^n)
  for (i in 1:(2^n)){
    T[i] <- sum(S[i,]*Xd)
  }
  
  #p-value
  p2 <- (sum(T >= abs(d)) + sum(T <= -abs(d)))/(2^n)
  return(p2)
  
}


conf_int <- function(X, alpha){
  #returns a 100*(1 - alpha) confidence interval for the location parameter
  #by inverting permutation tests

  m0s <- linspace(-2, 2, n=100)
  p2s = matrix(nrow=length(m0s))
  for (k in 1:length(m0s)){
    m0 <- m0s[k]
    
    p2s[k] <- perm_test2(X, m0)

  }
  
  #lower and upper bounds of confidence interval
  lb <- min(m0s[p2s >= alpha])
  ub <- max(m0s[p2s >= alpha])
  
  return(c(lb, ub))
}

#set random seed
set.seed(42)

#sample size
n <- 10
#level of significance
alpha <- 0.05
#actual location paraneter
theta <- 0

#generate random normal sample
X <- rnorm(n, mean=theta)

perm_test2(X, theta)
conf_int(X, alpha)